### PR TITLE
CTPPS: updated diamond unpacker for post-TS1 conditions

### PR DESCRIPTION
**Pull request by forthommel**
This PR introduces a collection of critical (and *urgent*) SW changes for 2017 data-taking operations with the diamond detectors:
* new "lite" readout mode unpacking handled properly, allowing to run on high rate conditions (previously limited at ~80 kHz) ; this SW update is fully backward-compatible, but its integration on the Tier-0 is **critical** before the deployment of the new FED FW
* leading and trailing edges are inverted, following the investigation performed on the signal cable polarities ; this feature will be backported in all useful releases for the next re-reco of previously collected data.

The test in `EventFilter/CTPPSRawToDigi/test/test_diamonds_only_cfg.py` has been modified to allow to test the new unpacker on a miniDAQ run collected with this new readout mode.

Updated data format is documented [here](https://docs.google.com/spreadsheets/d/1pNlddsWZDt1cnrVD0SCfPE9SgChksd3Mpbf0i0jvKLQ/edit#gid=0).

### CODE REVIEW

**Comment by perrotta** [comment]
paese test

**Comment by forthommel** [comment]
please test, you mean? :)

**Comment by smuzaffar** [comment]
please test

**Comment by perrotta** [comment]
please test
(Hard doing it from a cell phone with T9 set to Italian ;-) )

**Comment by forthommel** [comment]
The new version of this firmware with a lite data frame output is deployed on FEDs 582/3 since Friday. Therefore, the current Tier-0/central DQM release is currently skipping the unpacking of these ones (invalid readout mode).
You may hence find several higher-statistics runs to test this amended unpacker with, including ones with circulating beam (for instance, `299042`).

**Comment by slava77** [comment]
urgent

**Comment by forthommel** [comment]
Thanks @slava77 
Please let us know asap if you need any input from us to ease this review!

**Comment by slava77** [comment]
@forthommel 
please add a pointer to the documentation of the (updated) data format specifications to the description of this PR.

File: EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h [hunk 1]
```diff
@@ -28,14 +31,14 @@ class DiamondVFATFrame : public VFATFrame
     /// get timing infromation
     uint32_t getLeadingEdgeTime() const
     {
-      uint32_t time = ((data[5]&0x1f)<<16)+data[6];
+      uint32_t time = ((data[7]&0x1f)<<16)+data[8];
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: just curious: is this a data format change or a bugfix (that the leading and trailing were actually swapped)

File: EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h [hunk 2]
```diff
@@ -26,7 +27,8 @@ class RawDataUnpacker
     typedef uint64_t word;
 
     /// VFAT transmission modes
-    enum { vmCluster = 0x80, vmRaw = 0x90 };
+    enum { vmCluster = 0x80, vmRaw = 0x90, vmDiamondCompact = 0xB0 };
+    enum { VFAT_DIAMOND_HEADER_OF_WORD_2=0x7800, VFAT_DIAMOND_HEADER_OF_WORD_3=0x7000, VFAT_DIAMOND_HEADER_OF_WORD_5=0x6800, VFAT_DIAMOND_HEADER_OF_WORD_7=0x6000, VFAT_HEADER_OF_EC=0xC000 };
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h]: please add some linebreaks to improve readability.
Also, these are becoming long enough that it seems more appropriate to make these constexpr with explicit type specification.

**Comment by nminafra** [comment]
@slava77 The data formats (full and compact) are available here.
https://docs.google.com/spreadsheets/d/1pNlddsWZDt1cnrVD0SCfPE9SgChksd3Mpbf0i0jvKLQ/edit?usp=sharing
They are both based on the VFAT data frame to keep the code (and DAQ) as much similar as possible to CTPPS strip detectors.

File: EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h (hunk 1 above)
Comment (nminafra) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: It was actually an hardware bug. This will be fine for all 2017 data.

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 3]
```diff
@@ -80,11 +80,13 @@ int RawDataUnpacker::ProcessOptoRxFrame(const word *buf, unsigned int frameSize,
...
     return ProcessOptoRxFrameSerial(buf, frameSize, fc);
+  }
 
-  if (FOV == 2)
+  if (FOV == 2 || FOV == 3) {
     return ProcessOptoRxFrameParallel(buf, frameSize, fedInfo, fc);
+  }
 
   if (verbosity)
     LogWarning("Totem") << "Error in RawDataUnpacker::ProcessOptoRxFrame > " << "Unknown FOV = " << FOV << endl;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: it looks like the independent ifs starting from ```if (FOV == 1) ``` are better handled with a switch-case or with an if-else

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 4]
```diff
@@ -275,6 +277,14 @@ int RawDataUnpacker::ProcessVFATDataParallel(const uint16_t *buf, unsigned int O
 
   if (hFlag == vmRaw)
     wordsProcessed += 9;
+  
+  if (hFlag == vmDiamondCompact)
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: case or else-if for the vmCluster, vmRaw, vmDiamondCompact seems more appropriate

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 5]
```diff
@@ -275,6 +277,14 @@ int RawDataUnpacker::ProcessVFATDataParallel(const uint16_t *buf, unsigned int O
...
+  {
+    wordsProcessed--;
+    while ( (buf[wordsProcessed] & 0xFFF0)!= 0xF000 ) {
+      wordsProcessed++;
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: should there be some run-away protection?
(same for another while loop in vmCluster)

File: EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h (hunk 1 above)
Comment (davidlange6) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: how does 2016 get handled? (or is there none to worry about)

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 6]
```diff
@@ -374,6 +384,38 @@ int RawDataUnpacker::ProcessVFATDataParallel(const uint16_t *buf, unsigned int O
...
+      if ( ( buf[i] & 0xF000 ) == VFAT_HEADER_OF_EC ) {     // If Event Couter word is found
+        fd[10] = buf[i];
+        continue;
+      }
+      switch ( buf[i] & 0xF800 ) {
+        case VFAT_DIAMOND_HEADER_OF_WORD_2: // if Word 2 of the diamond VFAT frame is found
+          fd[2] = buf[i];
+          fd[1] = buf[i + 1];
+          break;
+        case VFAT_DIAMOND_HEADER_OF_WORD_3: // if Word 2 of the diamond VFAT frame is found
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: the commented text doesn't match the value checked

File: EventFilter/CTPPSRawToDigi/src/RawToDigiConverter.cc [hunk 7]
```diff
@@ -280,14 +282,15 @@ void RawToDigiConverter::Run(const VFATFrameCollection &coll, const TotemDAQMapp
     if (record.status.isOK())
     {
       const VFATFrame *fr = record.frame;
-      DiamondVFATFrame *diamondframe = (DiamondVFATFrame*) fr;
+      DiamondVFATFrame *diamondframeTmp = (DiamondVFATFrame*) fr;
+      DiamondVFATFrame diamondframe(*diamondframeTmp);
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawToDigiConverter.cc]: is the full copy really necessary?
The use case below suggests that it is not needed 
and that this file should have no modifications.

File: EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h (hunk 1 above)
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: there is 2016 data in 2016H period.
When was this bug fixed?

Comment (slava77) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: from the PR description it sounds like this swap has always been there and that all past data will need this fix.

**Comment by slava77** [comment]
@cmsbuild please test workflow 136.772,136.780

**Comment by slava77** [comment]
On 7/18/17 5:45 AM, Nicola Minafra wrote:
> @slava77 <https://github.com/slava77> The data formats (full and
> compact) are available here.
> https://docs.google.com/spreadsheets/d/1pNlddsWZDt1cnrVD0SCfPE9SgChksd3Mpbf0i0jvKLQ/edit?usp=sharing
> They are both based on the VFAT data frame to keep the code (and DAQ) as
> much similar as possible to CTPPS strip detectors.

Is this applicable to the "latest version"
or does this apply also to all data that was taken already?

I see that the leading and trailing edge assignments in this doc match
what is done in the code.
Does this mean that this polarity swap is now a part of the specs (a
problem becoming a feature)?

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 8]
```diff
@@ -80,11 +80,14 @@ int RawDataUnpacker::ProcessOptoRxFrame(const word *buf, unsigned int frameSize,
...
-  if (FOV == 1)
-    return ProcessOptoRxFrameSerial(buf, frameSize, fc);
-
-  if (FOV == 2)
-    return ProcessOptoRxFrameParallel(buf, frameSize, fedInfo, fc);
+  switch (FOV) {
+    case 1:
+      return ProcessOptoRxFrameSerial(buf, frameSize, fc);
+    case 2:
+    case 3:
```
Comment (slava77) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: what does FOV=3 correspond to?
I do not see any other places where the code uses it to do something.

If this is for a firmware version that can have the compact format?

Comment (nminafra) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: Exactly: it is an extension of the FOV==2, including the compact data format for the diamonds.
Since it is anyway a new firmware, the FOV was incremented.

**Comment by nminafra** [comment]
@slava77 @davidlange6 The inversion problem was supposed to be there due to a wrong cabling (strangely enough): the inversion was introduced in the software to compensate. Looking at the data we realized that this was not needed.
We realized this only now since we've been focusing on the tracking; now that we started analyzing also the timing, we spotted the inversion (looking at isolated bunches).
In summary: this fix is retroactive and will work perfectly fine for all data ever collected with CTPPS diamonds; it was a solution to a bug that was never there.

File: EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h (hunk 1 above)
Comment (forthommel) [inline EventFilter/CTPPSRawToDigi/interface/DiamondVFATFrame.h]: Indeed, this leading-trailing edges inversion was also present in 2016 data: the two following screenshots show the time over threshold &equiv; trailing - leading edge (ns) respectively before and after the patch for an arbitrary number of events in run 284029 from 2016H era for a loose selection: screenshot from 2017-07-18 16-41-24
screenshot from 2017-07-18 16-33-11

**Comment by slava77** [comment]
@forthommel 
please prepare a backport of the leading-trailing edge swap to 80X (only that part).

**Comment by forthommel** [comment]
Hi @slava77 
#19806 has been opened in `8_0_X` for this words inversion.

**Comment by slava77** [comment]
Here are some notes from tests in CMSSW_9_3_0_pre2

- 2016 data from matrix workflow 136.772 and 2017 data from matrix workflow 136.78  (200 events in each) both show somewhat expected behavior: changes start from ctppsDiamondRecHits t0 values and then propagate downstream to track distributions. The change in time values is in clear agreement with the flip from leading to trailing edge
wf136 78_ctppsdiamond_450_220cyl_hitsinplanes
wf136 78_ctppsdiamond_450_220cyl_tot
wf136 78_ctppsdiamond_450_220cyl_tracks

- 2017 data after the new FW is in place 299042 lumi 41 in JetHT PD show an expected appearance of digis and hits, but unlike in the earlier examples the number of hits is pretty small and there were no tracks reconstructed
    - file /store/data/Run2017B/JetHT/RAW/v1/000/299/042/00000/1046825C-D368-E711-A36B-02163E011DF9.root
wfjetht299042_ctppsdiamond_450_220cyl_hitsinplanes
wfjetht299042_ctppsdiamond_450_220cyl_tot

and some "raw" distributions plotting from root files with  FWlite 
(black is with this PR [despite what the legend says])
all_origvssign940_jetht299042c_ctppsdiamondrechitedmdetsetvector_ctppsdiamondrechits__rereco_obj__sets_data_gett

Is this kind of hit multiplicity and absence of local tracks expected in the run 299042 lumi 41?
If not, perhaps there are some problems in unpacking.

**Comment by slava77** [comment]
In case it is useful:
Looking at ``` CTPPSDiamondDigiedmDetSetVector_ctppsDiamondRawToDigi_TimingDiamond_``` branch, to get some input logically upstream from the hits:
- run 297227 test with 200 events has 1452 digis with either leading or trailing edge not zero and 679 with both leading and trailing not zero
- run 299042 test has 418 with either leading or trailing not zero and only 10 with both leading and trailing not zero

**Comment by nminafra** [comment]
@slava77 Thanks for the comparison. Just keep in mind that the detector is still in commissioning phase, therefore the hardware configurations is changing very often. The problem of incomplete events (missing leading or trailing edge) is still under investigation (indeed, we have a plot in the DQM), but we believe it is hardware related since if we select only isolated bunch, most of the events are complete.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19754	62c87b8
- unpacker updated to handle new (compact) channel data; diamond digi trailing and leading edges are swapped back to the correct order 
- jenkins tests pass [including workflows with diamonds data, even though without comparisons]
- local tests confirm expected behavior as described in https://github.com/cms-sw/cmssw/pull/19754#issuecomment-316203806

File: EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h [hunk 9]
```diff
@@ -26,7 +27,14 @@ class RawDataUnpacker
     typedef uint64_t word;
 
     /// VFAT transmission modes
-    enum { vmCluster = 0x80, vmRaw = 0x90 };
+    enum { vmCluster = 0x80, vmRaw = 0x90, vmDiamondCompact = 0xB0 };
```
Comment (davidlange6) [inline EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h]: hi @forthommel - i realize a class with such a generic name as RawDataUnpacker should be in a ctpps namespace. is that a big problem?

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 10]
```diff
@@ -374,6 +380,43 @@ int RawDataUnpacker::ProcessVFATDataParallel(const uint16_t *buf, unsigned int O
     presenceFlags |= 0x8;
     fd[0] = buf[dataOffset + 8];
   }
+  
+  // get channel data for diamond compact mode
+  if (hFlag == vmDiamondCompact)
+  {
+    for (unsigned int i = 1; (buf[i+1] & 0xFFF0)!= 0xF000; i++) {
```
Comment (davidlange6) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: should this not protect against an infinite loop? i guess by checking against maxwords ?

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc [hunk 11]
```diff
@@ -374,6 +380,43 @@ int RawDataUnpacker::ProcessVFATDataParallel(const uint16_t *buf, unsigned int O
...
+      if ( ( buf[i] & 0xF000 ) == VFAT_HEADER_OF_EC ) {
+        // Event Counter word is found
+        fd[10] = buf[i];
+        continue;
+      }
+      switch ( buf[i] & 0xF800 ) {
+        case VFAT_DIAMOND_HEADER_OF_WORD_2:
+          // word 2 of the diamond VFAT frame is found
+          fd[2] = buf[i];
+          fd[1] = buf[i + 1];
```
Comment (davidlange6) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: just confirming that this is not a typo - fd[2] followed by fd[1] looks like one - might be worth a comment

Comment (forthommel) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: It is not a typo, only a feature of our frame packing format. If that is any clearer I can swap the two lines, as it seems a new commit + test + (+1) round is needed.

File: EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h (hunk 9 above)
Comment (forthommel) [inline EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h]: Hi @davidlange6 
Sure, I will wrap this object in a `ctpps` namespace in my next iteration

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc (hunk 10 above)
Comment (forthommel) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: Addressed in 4831ff2

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Sorry @slava77, I protected another loop instead of the one @davidlange6 was referring to! Yet another commit to push, could you abort the test?

File: EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h (hunk 9 above)
Comment (Martin-Grunewald) [inline EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h]: Within ConfDB we can not have bla::foobar but only blafoobar as module plugin name.
So if this unpacker needs to run within HLT eventually, you'd need to make a unique name,
such as CTPPSRawDataUnpacker.

File: EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc (hunk 10 above)
Comment (forthommel) [inline EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc]: Loop protected in https://github.com/cms-sw/cmssw/pull/19754/commits/a2ba7c69467364715812ad7c65ebe2e284485dc9

**Comment by slava77** [comment]
@forthommel 
from the console link you can see that the tests are running from a2ba7c6

**Comment by slava77** [comment]
On 7/19/17 1:45 AM, Martin Grunewald wrote:
> *@Martin-Grunewald* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h
> <https://github.com/cms-sw/cmssw/pull/19754#discussion_r128187343>:
> 
>> @@ -26,7 +27,14 @@ class RawDataUnpacker
>      typedef uint64_t word;
>  
>      /// VFAT transmission modes
> -    enum { vmCluster = 0x80, vmRaw = 0x90 };
> +    enum { vmCluster = 0x80, vmRaw = 0x90, vmDiamondCompact = 0xB0 };
> 
> Within ConfDB we can not have bla::foobar but only blafoobar as module
> plugin name.
> So if this unpacker needs to run within HLT eventually, you'd need to
> make a unique name,
> such as CTPPSRawDataUnpacker.

this is not a plugin.
you will not see this in python

**Comment by forthommel** [comment]
Perfect @slava77 ! Thanks a lot!

**Comment by davidlange6** [comment]
ah - good to know.. CTPPSRawDataUnpacker would also follow the pattern of other unpackers in CMSSW

**Comment by forthommel** [comment]
OK, let me know what you find the most "CMSSWable". My prior experience was with the tracker strips, where a namespace is also used (see [here](https://github.com/cms-sw/cmssw/blob/master/EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc#L22) for instance)

**Comment by Martin-Grunewald** [comment]
What counts is the plugin name used in the FW maker macro ```DEFINE_FWK_MODULE```
https://github.com/cms-sw/cmssw/blob/master/EventFilter/SiStripRawToDigi/plugins/module.cc
see typedefs there.

**Comment by Martin-Grunewald** [comment]
The longer you wait the harder it gets as you need to fix all usage creeping in.
But as I said, it is needed only in case you (urgently?) want to have this at HLT,
and the we'd need a 92X backport.

**Comment by davidlange6** [comment]
please test

**Comment by forthommel** [comment]
@davidlange6, as mentioned by @slava77 in https://github.com/cms-sw/cmssw/pull/19754#issuecomment-316317452 the previous batch of tests were performed on top of https://github.com/cms-sw/cmssw/commit/a2ba7c69467364715812ad7c65ebe2e284485dc9.

**Comment by jan-kaspar** [comment]
> What counts is the plugin name used in the FW maker macro DEFINE_FWK_MODULE
> https://github.com/cms-sw/cmssw/blob/master/EventFilter/SiStripRawToDigi/plugins/module.cc
> see typedefs there.

Our raw-to-digi plugin already carries a prefix which should avoid any confusion:
https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/plugins/TotemVFATRawToDigi.cc#L186

The "RawDataUnpacker"
https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h
is internal code only, not a FW module, thus I can't see the problem.

**Comment by davidlange6** [comment]
> On Jul 19, 2017, at 12:29 PM, jan-kaspar <notifications@github.com> wrote:
> 
> What counts is the plugin name used in the FW maker macro DEFINE_FWK_MODULE
> https://github.com/cms-sw/cmssw/blob/master/EventFilter/SiStripRawToDigi/plugins/module.cc
> see typedefs there.
> 
> Our raw-to-digi plugin already carries a prefix which should avoid any confusion:
> https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/plugins/TotemVFATRawToDigi.cc#L186
> 
> The "RawDataUnpacker"
> https://github.com/cms-sw/cmssw/blob/master/EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h
> is internal code only, not a FW module, thus I can't see the problem.
> 

in what sense is it "internal" Are its symbols hidden from the rest of CMSSW?

**Comment by jan-kaspar** [comment]
> in what sense is it "internal" Are its symbols hidden from the rest of CMSSW?

I'm not quite sure how scram handles this. The "RawDataUnpacker" is a class defined in EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h, implemented in EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc and only used by code in EventFilter/CTPPSRawToDigi/plugins . So no other package should have EventFilter/CTPPSRawToDigi in its BuildFile. Is this a safe configuration?

**Comment by davidlange6** [comment]
> On Jul 19, 2017, at 12:48 PM, jan-kaspar <notifications@github.com> wrote:
> 
> in what sense is it "internal" Are its symbols hidden from the rest of CMSSW?
> 
> I'm not quite sure how scram handles this. The "RawDataUnpacker" is a class defined in EventFilter/CTPPSRawToDigi/interface/RawDataUnpacker.h, implemented in EventFilter/CTPPSRawToDigi/src/RawDataUnpacker.cc and only used by code in EventFilter/CTPPSRawToDigi/plugins . So no other package should have EventFilter/CTPPSRawToDigi in its BuildFile. Is this a safe configuration?
> 

No -
1) any package is allowed (by policy anyway) to load things in libraries. If you want to prevent that, put your code together with your plugins.
2) at the end of the day, all libraries and symbols get loaded into one application. So conflicts in namespace are there regardless...

**Comment by jan-kaspar** [comment]
> >  Is this a safe configuration?

> No -

OK, then the easiest might be to introduce a namespace for what is in interface and src. plugins directory doesn't need it as the names are unambiguous (CTPPS or Totem prefix) - this also avoids troubles with ConfDB as far as I understand. What do you think?

**Comment by davidlange6** [comment]
i think that is what has happened.

**Comment by slava77** [comment]
On 7/19/17 4:33 AM, David Lange wrote:
> i think that is what has happened.

moving to ctpps namespace was already done in this PR.
Where is the remaining problem?

(confDB case is irrelevant because the code is not a FW plugin and is
not planned to be)


>

**Comment by jan-kaspar** [comment]
> moving to ctpps namespace was already done in this PR.
> Where is the remaining problem?

There is absolutely no problem, I apologise for my outdated comments (I got lost in the thread).

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19754	a2ba7c6
- mostly technical updates since the last signoff https://github.com/cms-sw/cmssw/pull/19754/files/62c87b83fb0b96ce7b6fed45bde2b926fe6c2045..a2ba7c69467364715812ad7c65ebe2e284485dc9?w=1
    - namespace ctpps added to somewhat generically named code to avoid symbol conflicts
    - while and for loops over the input data buffers are protected from runaway conditions (e.g. from corrupted data
- jenkins tests pass

**Comment by forthommel** [comment]
@davidlange6 is there any other concern on your side?

**Comment by davidlange6** [comment]
+1

**Comment by slava77** [comment]
@forthommel 
it looks like it's time for you to make a back-port for 92X
